### DISCUSSION TITLE
Switch state as MQTT message

### DISCUSSION THREAD

**Discussion by froido** [arendst/Tasmota#15369] [Q&A] [answered]
Hi,
I am looking for an option doing as follows
1. when switch1 is toggled relais1 (power1) shall be toggled (switchmode 0), to minimize the power toggle delay
2. when switch1 is toggled an MQTT message shall be published (like switchmode 15)

My goal is to get knowledge if the relais was switched by an MQTT command (external, like OpenHab) or if the relais was toggled by the connected physical switch. Therefore I want to keep the direct link between switch and relais, but get an MQTT message about the switch state. 

Do you have any idea how to realise this?

**Comment by ascillato** [comment, marked as answer]
Hi,

> 1. when switch1 is toggled relais1 (power1) shall be toggled (switchmode 0), to minimize the power toggle delay 

Please, take a look at `SetOption13` in the docs at https://tasmota.github.io/docs/Commands/#setoptions

> 2. when switch1 is toggled an MQTT message shall be published (like switchmode 15)

Yes, you can use Switchmode15 and a rule if you want to control your relays, or the other options explained in the docs at https://tasmota.github.io/docs/Buttons-and-Switches/

> My goal is to get knowledge if the relais was switched by an MQTT command (external, like OpenHab) or if the relais was toggled by the connected physical switch.

In the console, if you set `weblog 3` you will see **SRC:** that is where the command comes from, like physical Switch, MQTT command, etc.

**Reply by froido** [reply]
Thanks for the quick response.

> Please, take a look at SetOption13 in the docs at https://tasmota.github.io/docs/Commands/#setoptions

As I use a physical switch (not a button), SetOption13 would make no difference. Would it?

> Yes, you can use Switchmode15 and a rule if you want to control your relays, or the other options explained in the docs at https://tasmota.github.io/docs/Buttons-and-Switches/

This is an option I thought about, but it seems very uncomfortable as I want it to use on multiple tasmota devices. Maybe this could be some new feature implemented as an additional SetOption.

> In the console, if you set weblog 3 you will see SRC: that is where the command comes from, like physical Switch, MQTT command, etc.

Pushing the physical switch it shows MQTT as SRC, but I don't know why. SwitchMode1 is 0 and SetOption114 is OFF.

```
21:43:56.460 SRC: MQTT
21:43:56.463 CMD: Grp 0, Cmnd 'POWER', Idx 1, Len 3, Data 'OFF'
21:43:56.468 MQT: tasmota/shelly_1pm_flur/stat/RESULT = {"POWER":"OFF"}
21:43:56.473 MQT: tasmota/shelly_1pm_flur/stat/POWER = OFF

**Reply by ascillato2** [reply]
Please, show the previous part of that MQTT message. Seems that you have switchtopic defined.

**Reply by froido** [reply]
> Please, show the previous part of that MQTT message. Seems that you have switchtopic defined.

You are right, there was a switchtopic defined. After deleting everything was right. Now I see `SRC: Switch`.

> This is an option I thought about, but it seems very uncomfortable as I want it to use on multiple tasmota devices. Maybe this could be some new feature implemented as an additional SetOption.

How do you think aboud adding this as a feature?
